src/gpu/d3d11/SDL_gpu_d3d11.c

```diff
@@ -4131,6 +4501,147 @@ static void D3D11_BindIndexBuffer(
 	);
 }
 
+static void D3D11_BindGraphicsResourceSet(
+    SDL_GpuCommandBuffer *commandBuffer,
+    Uint32 setIndex,
+    SDL_GpuShaderResourceBinding *resourceBindings,
+    Uint32 resourceBindingCount
+) {
+    D3D11_INTERNAL_BindResourceSet(
+        commandBuffer,
+        SDL_TRUE,
+        setIndex,
+        resourceBindings,
+        resourceBindingCount
+    );
+}
+
+static void D3D11_DrawInstancedPrimitives(
+	SDL_GpuCommandBuffer *commandBuffer,
+	Uint32 baseVertex,
+	Uint32 startIndex,
+	Uint32 primitiveCount,
+	Uint32 instanceCount
+) {
+	D3D11CommandBuffer *d3d11CommandBuffer = (D3D11CommandBuffer*) commandBuffer;
+
+	ID3D11DeviceContext_DrawIndexedInstanced(
+		d3d11CommandBuffer->context,
+		PrimitiveVerts(d3d11CommandBuffer->graphicsPipeline->primitiveType, primitiveCount),
+		instanceCount,
+		startIndex,
+		baseVertex,
+		0
+	);
+}
+
+static void D3D11_DrawPrimitives(
+	SDL_GpuCommandBuffer *commandBuffer,
+	Uint32 vertexStart,
+	Uint32 primitiveCount
+) {
+	D3D11CommandBuffer *d3d11CommandBuffer = (D3D11CommandBuffer*) commandBuffer;
+
+	ID3D11DeviceContext_Draw(
+		d3d11CommandBuffer->context,
+		PrimitiveVerts(d3d11CommandBuffer->graphicsPipeline->primitiveType, primitiveCount),
+		vertexStart
+	);
+}
+
+static void D3D11_DrawPrimitivesIndirect(
+	SDL_GpuCommandBuffer *commandBuffer,
+	SDL_GpuBuffer *gpuBuffer,
+	Uint32 offsetInBytes,
+	Uint32 drawCount,
+	Uint32 stride
+) {
+	D3D11CommandBuffer *d3d11CommandBuffer = (D3D11CommandBuffer*) commandBuffer;
+	D3D11Buffer *d3d11Buffer = ((D3D11BufferContainer*) gpuBuffer)->activeBuffer;
+
+	/* D3D11: "We have multi-draw at home!"
+	 * Multi-draw at home:
+	 */
+	for (Uint32 i = 0; i < drawCount; i += 1)
+	{
+		ID3D11DeviceContext_DrawInstancedIndirect(
+			d3d11CommandBuffer->context,
+			d3d11Buffer->handle,
+			offsetInBytes + (stride * i)
+		);
+	}
+
+	D3D11_INTERNAL_TrackBuffer(d3d11CommandBuffer, d3d11Buffer);
+}
+
+static void D3D11_EndRenderPass(
+	SDL_GpuCommandBuffer *commandBuffer
+) {
+	D3D11CommandBuffer *d3d11CommandBuffer = (D3D11CommandBuffer*) commandBuffer;
+    Uint32 i;
+
+	/* Resolve MSAA color render targets */
+	for (i = 0; i < MAX_COLOR_TARGET_BINDINGS; i += 1)
+	{
+		if (d3d11CommandBuffer->colorTargetMsaaHandle[i] != NULL)
+		{
+			ID3D11DeviceContext_ResolveSubresource(
+				d3d11CommandBuffer->context,
+				d3d11CommandBuffer->colorTargetResolveTexture[i]->handle,
+				d3d11CommandBuffer->colorTargetResolveSubresourceIndex[i],
+				d3d11CommandBuffer->colorTargetMsaaHandle[i],
+				0,
+				SDLToD3D11_TextureFormat[d3d11CommandBuffer->colorTargetResolveTexture[i]->format]
+			);
+		}
+	}
+
+    /* Reset slots */
+    /* FIXME: review max slot count */
+    ID3D11DeviceContext_VSSetShaderResources(
+        d3d11CommandBuffer->context,
+        0,
+        MAX_VERTEXTEXTURE_SAMPLERS,
```

Comment:
it would be nice if there was some sizeof() stuff happening to make sure the arrays aren't too small